File: frontend/src/api/adguard.ts
import { api } from './client'

// null = master, number = slave instance ID
export type AdguardSource = null | number

export interface AdguardStatus {
  running: boolean
  version: string
  dns_addresses: string[]
  dns_port: number
  querylog_enabled: boolean
  protection_enabled: boolean
}

export interface SafeSearchSettings {
  enabled: boolean
  bing: boolean
  duckduckgo: boolean
  ecosia: boolean
  google: boolean
  pixabay: boolean
  yandex: boolean
  youtube: boolean
}

export interface AdguardStats {
  num_dns_queries: number
  num_blocked_filtering: number
  num_replaced_safebrowsing: number
  num_replaced_parental: number
  avg_processing_time: number
}

export interface AdguardFilter {
  id: number
  url: string
  name: string
  enabled: boolean
  rules_count: number
}

export interface AdguardFilteringStatus {
  enabled: boolean
  interval: number
  filters: AdguardFilter[]
}

export interface AdguardRewrite {
  domain: string
  answer: string
}

export interface AdguardServiceInfo {
  id: string
  name: string
  icon_svg: string
}

export interface AdguardServicesData {
  services: AdguardServiceInfo[]
  blocked: string[]
}

// ── Unified call ──────────────────────────────────────────────────────────────

async function call<T>(
  source: AdguardSource,
  method: string,
  path: string,
  body?: unknown,
): Promise<T> {
  if (source === null) {
    // Master — direct API call
    const adguardPath = '/adguard' + path
    switch (method.toUpperCase()) {
      case 'GET':    return api.get<T>(adguardPath).then((r) => r.data)
      case 'POST':   return api.post<T>(adguardPath, body).then((r) => r.data)
      case 'PUT':    return api.put<T>(adguardPath, body).then((r) => r.data)
      case 'DELETE': return api.delete<T>(adguardPath, { data: body }).then((r) => r.data)
      default:       throw new Error(`Unsupported method: ${method}`)
    }
  } else {
    // Slave — proxy through master
    const res = await api.post(`/instances/${source}/adguard/proxy`, {
      method,
      path: '/api/v1/adguard' + path,
      body: body ?? null,
    })
    const proxy = res.data as { status: number; body: string }
    if (proxy.status >= 400) {
      let msg = `AdGuard returned ${proxy.status}`
      try { msg = JSON.parse(proxy.body)?.error ?? msg } catch { /* noop */ }
      throw new Error(msg)
    }
    if (!proxy.body || proxy.body === '') return undefined as T
    return JSON.parse(proxy.body) as T
  }
}

// ── API functions ─────────────────────────────────────────────────────────────

export const getStatus = (src: AdguardSource) =>
  call<AdguardStatus>(src, 'GET', '/status')

export const getStats = (src: AdguardSource) =>
  call<AdguardStats>(src, 'GET', '/stats')

export const getFilteringStatus = (src: AdguardSource) =>
  call<AdguardFilteringStatus>(src, 'GET', '/filtering')

export const setFilteringEnabled = (src: AdguardSource, enabled: boolean) =>
  call<void>(src, 'PUT', '/filtering/config', { enabled, interval: 24 })

export const addFilter = (src: AdguardSource, url: string, name: string) =>
  call<void>(src, 'POST', '/filtering/add', { url, name })

export const removeFilter = (src: AdguardSource, url: string) =>
  call<void>(src, 'POST', '/filtering/remove', { url })

export const refreshFilters = (src: AdguardSource) =>
  call<void>(src, 'POST', '/filtering/refresh', {})

export const getUserRules = (src: AdguardSource) =>
  call<{ rules: string[] }>(src, 'GET', '/rules')

export const setUserRules = (src: AdguardSource, rules: string[]) =>
  call<void>(src, 'PUT', '/rules', { rules })

export const getRewrites = (src: AdguardSource) =>
  call<{ rewrites: AdguardRewrite[] }>(src, 'GET', '/rewrites')

export const addRewrite = (src: AdguardSource, domain: string, answer: string) =>
  call<void>(src, 'POST', '/rewrites', { domain, answer })

export const deleteRewrite = (src: AdguardSource, domain: string, answer: string) =>
  call<void>(src, 'DELETE', '/rewrites', { domain, answer })

export const setProtection = (src: AdguardSource, enabled: boolean) =>
  call<void>(src, 'POST', '/protection', { enabled })

export const getSafeBrowsingStatus = (src: AdguardSource) =>
  call<{ enabled: boolean }>(src, 'GET', '/safebrowsing')

export const setSafeBrowsing = (src: AdguardSource, enabled: boolean) =>
  call<void>(src, 'PUT', '/safebrowsing', { enabled })

export const getParentalStatus = (src: AdguardSource) =>
  call<{ enabled: boolean }>(src, 'GET', '/parental')

export const setParental = (src: AdguardSource, enabled: boolean) =>
  call<void>(src, 'PUT', '/parental', { enabled })

export const getSafeSearchStatus = (src: AdguardSource) =>
  call<SafeSearchSettings>(src, 'GET', '/safesearch')

export const setSafeSearch = (src: AdguardSource, settings: SafeSearchSettings) =>
  call<void>(src, 'PUT', '/safesearch', settings)

export const getServices = (src: AdguardSource) =>
  call<AdguardServicesData>(src, 'GET', '/services')

export const setBlockedServices = (src: AdguardSource, ids: string[]) =>
  call<void>(src, 'PUT', '/services', { ids })

File: backend/internal/api/handlers/adguard.go
package handlers

import (
	"net/http"

	"github.com/AlexArtaud-Dev/velar/backend/internal/services/adguard"
	"github.com/gin-gonic/gin"
)

type AdguardHandler struct {
	ag *adguard.Client
}

func NewAdguardHandler(ag *adguard.Client) *AdguardHandler {
	return &AdguardHandler{ag: ag}
}

func (h *AdguardHandler) GetStatus(c *gin.Context) {
	status, err := h.ag.GetStatus()
	if err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"error": err.Error(), "running": false})
		return
	}
	c.JSON(http.StatusOK, status)
}

func (h *AdguardHandler) GetStats(c *gin.Context) {
	stats, err := h.ag.GetStats()
	if err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, stats)
}

func (h *AdguardHandler) GetFilteringStatus(c *gin.Context) {
	status, err := h.ag.GetFilteringStatus()
	if err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, status)
}

func (h *AdguardHandler) SetFilteringConfig(c *gin.Context) {
	var req adguard.FilteringConfig
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if err := h.ag.SetFilteringConfig(req); err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"error": err.Error()})
		return
	}
	c.Status(http.StatusNoContent)
}

func (h *AdguardHandler) AddFilter(c *gin.Context) {
	var req struct {
		URL  string `json:"url" binding:"required"`
		Name string `json:"name" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if err := h.ag.AddFilter(req.URL, req.Name); err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"error": err.Error()})
		return
	}
	c.Status(http.StatusCreated)
}

func (h *AdguardHandler) RemoveFilter(c *gin.Context) {
	var req struct {
		URL string `json:"url" binding:"required"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if err := h.ag.RemoveFilter(req.URL); err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"error": err.Error()})
		return
	}
	c.Status(http.StatusNoContent)
}

func (h *AdguardHandler) RefreshFilters(c *gin.Context) {
	if err := h.ag.RefreshFilters(); err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"error": err.Error()})
		return
	}
	c.Status(http.StatusNoContent)
}

func (h *AdguardHandler) GetUserRules(c *gin.Context) {
	rules, err := h.ag.GetUserRules()
	if err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"rules": rules})
}

func (h *AdguardHandler) SetUserRules(c *gin.Context) {
	var req struct {
		Rules []string `json:"rules"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if req.Rules == nil {
		req.Rules = []string{}
	}
	if err := h.ag.SetUserRules(req.Rules); err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"error": err.Error()})
		return
	}
	c.Status(http.StatusNoContent)
}

func (h *AdguardHandler) GetRewrites(c *gin.Context) {
	rewrites, err := h.ag.GetDNSRewrites()
	if err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"rewrites": rewrites})
}

func (h *AdguardHandler) AddRewrite(c *gin.Context) {
	var req adguard.DNSRewrite
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if req.Domain == "" || req.Answer == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "domain and answer are required"})
		return
	}
	if err := h.ag.AddDNSRewrite(req.Domain, req.Answer); err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"error": err.Error()})
		return
	}
	c.Status(http.StatusCreated)
}

func (h *AdguardHandler) DeleteRewrite(c *gin.Context) {
	var req adguard.DNSRewrite
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if err := h.ag.DeleteDNSRewrite(req.Domain, req.Answer); err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"error": err.Error()})
		return
	}
	c.Status(http.StatusNoContent)
}

func (h *AdguardHandler) SetProtection(c *gin.Context) {
	var req struct {
		Enabled bool `json:"enabled"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if err := h.ag.SetProtection(req.Enabled); err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"error": err.Error()})
		return
	}
	c.Status(http.StatusNoContent)
}

func (h *AdguardHandler) GetSafeBrowsingStatus(c *gin.Context) {
	enabled, err := h.ag.GetSafeBrowsingStatus()
	if err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"enabled": enabled})
}

func (h *AdguardHandler) SetSafeBrowsing(c *gin.Context) {
	var req struct{ Enabled bool `json:"enabled"` }
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if err := h.ag.SetSafeBrowsing(req.Enabled); err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"error": err.Error()})
		return
	}
	c.Status(http.StatusNoContent)
}

func (h *AdguardHandler) GetParentalStatus(c *gin.Context) {
	enabled, err := h.ag.GetParentalStatus()
	if err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"enabled": enabled})
}

func (h *AdguardHandler) SetParental(c *gin.Context) {
	var req struct{ Enabled bool `json:"enabled"` }
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if err := h.ag.SetParental(req.Enabled); err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"error": err.Error()})
		return
	}
	c.Status(http.StatusNoContent)
}

func (h *AdguardHandler) GetSafeSearchStatus(c *gin.Context) {
	settings, err := h.ag.GetSafeSearchStatus()
	if err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, settings)
}

func (h *AdguardHandler) SetSafeSearch(c *gin.Context) {
	var req adguard.SafeSearchSettings
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if err := h.ag.SetSafeSearch(req); err != nil {
		c.JSON(http.StatusBadGateway, gin.H{"error": err.Error()})
		return
	}
	c.Status(http.StatusNoContent)
}
